File: exercise/python/day13/model/shopping_sleigh.py
from model.discount import Discount
from model.special_offer_type import SpecialOfferType


class ShoppingSleigh:
    def __init__(self):
        self.items = []
        self.product_quantities = {}

    def get_items(self):
        return self.items

    def add_item(self, product):
        self.add_item_quantity(product, 1.0)

    def add_item_quantity(self, product, quantity):
        self.items.append((product, quantity))
        if product in self.product_quantities:
            self.product_quantities[product] += quantity
        else:
            self.product_quantities[product] = quantity

    def get_product_quantities(self):
        return self.product_quantities.copy()

    def handle_offers(self, receipt, offers, catalog):
        for product, quantity in self.product_quantities.items():
            if product in offers:
                offer = offers[product]
                unit_price = catalog.get_unit_price(product)
                quantity_as_int = int(quantity)
                discount = None
                x = 1

                if offer.offer_type == SpecialOfferType.THREE_FOR_TWO:
                    x = 3
                elif offer.offer_type == SpecialOfferType.TWO_FOR_AMOUNT:
                    x = 2
                    if quantity_as_int >= 2:
                        total = offer.argument * (quantity_as_int // x) + (quantity_as_int % 2) * unit_price
                        discount_amount = unit_price * quantity - total
                        discount = Discount(product, f"2 for {offer.argument}", -discount_amount)

                if offer.offer_type == SpecialOfferType.FIVE_FOR_AMOUNT:
                    x = 5

                number_of_xs = int(quantity_as_int / x)

                if offer.offer_type == SpecialOfferType.THREE_FOR_TWO and quantity_as_int > 2:
                    discount_amount = quantity * unit_price - ((number_of_xs * 2 * unit_price) + (quantity_as_int % 3) * unit_price)
                    discount = Discount(product, "3 for 2", -discount_amount)

                if offer.offer_type == SpecialOfferType.TEN_PERCENT_DISCOUNT:
                    discount = Discount(product, f"{offer.argument}% off", -quantity * unit_price * offer.argument / 100.0)

                if offer.offer_type == SpecialOfferType.FIVE_FOR_AMOUNT and quantity_as_int >= 5:
                    discount_amount = unit_price * quantity - (offer.argument * number_of_xs + (quantity_as_int % 5) * unit_price)
                    discount = Discount(product, f"5 for {offer.argument}", -discount_amount)

                if discount is not None:
                    receipt.add_discount(discount)

File: solution/day13/src/main/java/santamarket/model/ShoppingSleigh.java
package santamarket.model;

import java.util.*;

public class ShoppingSleigh {
    private final List<ProductQuantity> items = new ArrayList<>();
    private final Map<Product, Double> productQuantities = new HashMap<>();

    List<ProductQuantity> getItems() {
        return Collections.unmodifiableList(items);
    }

    void addItem(Product product) {
        addItemQuantity(product, 1.0);
    }

    Map<Product, Double> productQuantities() {
        return Collections.unmodifiableMap(productQuantities);
    }

    public void addItemQuantity(Product product, double quantity) {
        items.add(new ProductQuantity(product, quantity));
        if (productQuantities.containsKey(product)) {
            productQuantities.put(product, productQuantities.get(product) + quantity);
        } else {
            productQuantities.put(product, quantity);
        }
    }

    void handleOffers(Receipt receipt, Map<Product, Offer> offers, SantamarketCatalog catalog) {
        for (Product p : productQuantities().keySet()) {
            double quantity = productQuantities.get(p);
            if (offers.containsKey(p)) {
                Offer offer = offers.get(p);
                Discount discount = null;

                if (offer.offerType == SpecialOfferType.TWO_FOR_ONE) {
                    discount = computeXForYDiscount(new XForYOffer(2, 1), p, (int) quantity, catalog.getUnitPrice(p));
                } else if (offer.offerType == SpecialOfferType.THREE_FOR_TWO) {
                    discount = computeXForYDiscount(new XForYOffer(3, 2), p, (int) quantity, catalog.getUnitPrice(p));
                } else {
                    double unitPrice = catalog.getUnitPrice(p);
                    int quantityAsInt = (int) quantity;

                    int x = 1;
                    if (offer.offerType == SpecialOfferType.TWO_FOR_AMOUNT) {
                        x = 2;
                        if (quantityAsInt >= 2) {
                            double total = offer.argument * ((double) quantityAsInt / x) + quantityAsInt % 2 * unitPrice;
                            double discountN = unitPrice * quantity - total;
                            discount = new Discount(p, "2 for " + offer.argument, -discountN);
                        }
                    }
                    if (offer.offerType == SpecialOfferType.FIVE_FOR_AMOUNT) {
                        x = 5;
                    }
                    int numberOfXs = quantityAsInt / x;
                    if (offer.offerType == SpecialOfferType.TEN_PERCENT_DISCOUNT) {
                        discount = new Discount(p, offer.argument + "% off", -quantity * unitPrice * offer.argument / 100.0);
                    }
                    if (offer.offerType == SpecialOfferType.FIVE_FOR_AMOUNT && quantityAsInt >= 5) {
                        double discountTotal = unitPrice * quantity - (offer.argument * numberOfXs + quantityAsInt % 5 * unitPrice);
                        discount = new Discount(p, x + " for " + offer.argument, -discountTotal);
                    }
                }
                if (discount != null)
                    receipt.addDiscount(discount);
            }
        }
    }

    private Discount computeXForYDiscount(XForYOffer offer, Product p, int quantity, double unitPrice) {
        if (quantity <= offer.y())
            return null;
        int numberOfXs = quantity / offer.x();
        double discountAmount = unitPrice * (quantity - numberOfXs * offer.y() - quantity % offer.x());
        return new Discount(p, offer.x() + " for " + offer.y(), -discountAmount);
    }
}